elki/src/main/java/de/lmu/ifi/dbs/elki/math/statistics/dependence/McdeMwpDependenceMeasure.java

```diff
@@ -0,0 +1,227 @@
+     * @param adapter ELKI NumberArrayAdapter Subclass
+     * @param data One dimensional array containing one dimension of the data
+     * @param idx Return value of sortedIndex()
+     * @return Array of doubles, 3 subsequent values being assigned to one data instance.
+     * Containing sorted (ascending) row numbers, adjusted ranks and tying value corrections
+     * as required by MWP test. Example:
+     * double[] corrected_ranks = corrected_ranks(...);
+     * double l = corrected_rank[0]; double adjusted_rank = corrected_rank[1]; double correction = corrected_rank[2];
+     * 
+     * Correspond to one instance of the original data
+     */
+    protected <A> MwpIndex[] corrected_ranks(final NumberArrayAdapter<?, A> adapter, final A data, int[] idx){
+        final int len = adapter.size(data);
+        MwpIndex[] I = (MwpIndex[]) Array.newInstance(MwpIndex.class, len);
+
+        int j = 0; int correction = 0;
+        while(j < len){
+            int k = j; int t = 1; double adjust = 0.0;
+
+            while((k < len - 1) && (adapter.getDouble(data, idx[k]) == adapter.getDouble(data, idx[k+1]))){
+                adjust += k;
+                k++; t++;
+            }
+
+            if(k > j){
+                double adjusted = (adjust + k) / t;
+                correction += (t*t*t) - t;
+
+                for(int m = j; m <= k; m++){
+                    I[m] = new MwpIndex(idx[m], adjusted, correction);
+                }
+            }
+            else
+                I[j] = new MwpIndex(idx[j], j, correction);
+            j += t;
+        }
+
+        return I;
+    }
+
+    /**
+     * Efficient implementation of MWP statistical test using appropriate index structure as described in Algorithm 3
+     * of reference paper.
+     *
+     * @param len No of data instances
+     * @param slice Return value of randomSlice() created with the index that is not for the reference dimension
+     * @param corrected_ranks Index of the reference dimension, return value of corrected_ranks() computed for reference dimension
+     * @return p-value from two sided Mann-Whitney-U test
+     */
+    protected double statistical_test(int len, boolean[] slice, MwpIndex[] corrected_ranks){
+        final Random random = rnd.getSingleThreadedRandom(); // Note: No "safecut".
+        final int start = random.nextInt((int) (len * (1 - this.beta)));
+        final int end = start + (int) Math.ceil(len * this.beta);
+
+        double R = 0.0; long n1 = 0;
+        for(int j = start; j < end; j++){
+
+            if(slice[corrected_ranks[j].index]){
+                R += corrected_ranks[j].adjusted;
+                n1++;
+            }
+        }
+
+        // This is to cancel the offset in case the marginal restriction does not start from 0
+        // see "acc - (cutStart * count)" is reference implementation of MWP
+        R -= start * n1;
+
+        final int cutLength = end - start;
+        if((n1 == 0) || (n1 == cutLength)) return 1;
```

Comment:
By the ELKI code style (please also use the provided Eclipse code formatter in the `/conventions` folder to have our usual indentation of 2 spaces), please do *not* omit the `{}` parentheses. It is a best practise to always use parentheses even with single-line statements to avoid programming errors, such as the infamous Apple "goto fail" bug, which would have been prevented by not omitting the braces (and, of course, reachability analysis).